File: packages/pdf-gen-py/pdf_gen/generator.py
from typing import Callable, Dict, Any
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import A4, Letter, A3, A5
from reportlab.lib.units import inch
from io import BytesIO

from .config import PDFConfig, PAGE_SIZES

class PDFGenerator:
    def __init__(self, config: PDFConfig = None):
        if config is None:
            config = PDFConfig()
        self.config = config
        self.templates: Dict[str, Callable] = {}
        self.output = BytesIO()
        self.canvas = None
        self._setup_canvas()

    def _setup_canvas(self):
        page_sizes = {
            'A4': A4,
            'Letter': Letter,
            'A3': A3,
            'A5': A5,
        }

        page_size = page_sizes.get(self.config.page_size, A4)
        if self.config.orientation == 'landscape':
            page_size = (page_size[1], page_size[0])

        self.canvas = canvas.Canvas(self.output, pagesize=page_size)

    def register_template(self, name: str, renderer: Callable) -> None:
        self.templates[name] = renderer

    def use_template(self, template_name: str, data: Dict[str, Any]) -> None:
        if template_name not in self.templates:
            available = ', '.join(self.templates.keys())
            raise ValueError(f'Template "{template_name}" not found. Available: {available}')

        template = self.templates[template_name]
        template(self.canvas, self.config, data)

    def add_text(self, text: str, x: float = None, y: float = None, **options) -> 'PDFGenerator':
        if x is None:
            x = self.config.margins['left']
        if y is None:
            y = self.config.margins['top']

        self.canvas.drawString(x, y, text)
        return self

    def add_image(self, image_path: str, x: float, y: float, width: float = None, height: float = None) -> 'PDFGenerator':
        self.canvas.drawImage(image_path, x, y, width=width, height=height)
        return self

    def add_table(self, columns: list, rows: list, x: float = None, y: float = None, **options) -> 'PDFGenerator':
        if x is None:
            x = self.config.margins['left']
        if y is None:
            y = self.config.margins['top']

        page_width = 595.28 - self.config.margins['left'] - self.config.margins['right']
        col_width = page_width / len(columns)
        row_height = options.get('row_height', 20)

        # Draw header
        self.canvas.setFillColor('#f0f0f0')
        for i, col in enumerate(columns):
            self.canvas.rect(x + i * col_width, y, col_width, row_height, fill=True)
            self.canvas.drawString(x + i * col_width + 5, y + 8, str(col))

        # Draw rows
        for row_idx, row in enumerate(rows):
            for col_idx, cell in enumerate(row):
                self.canvas.rect(x + col_idx * col_width, y + (row_idx + 1) * row_height, col_width, row_height)
                self.canvas.drawString(x + col_idx * col_width + 5, y + (row_idx + 1) * row_height + 8, str(cell))

        return self

    def add_page(self) -> 'PDFGenerator':
        self.canvas.showPage()
        return self

    def add_line(self, x1: float, y1: float, x2: float, y2: float, color: str = 'black') -> 'PDFGenerator':
        self.canvas.line(x1, y1, x2, y2)
        return self

    def set_font(self, font_name: str, size: int) -> 'PDFGenerator':
        self.canvas.setFont(font_name, size)
        return self

    def generate(self, filename: str) -> None:
        self.canvas.save()
        with open(filename, 'wb') as f:
            f.write(self.output.getvalue())

File: packages/pdf-gen-js/src/generator.ts
import fs from 'fs';
import { PDFConfig, DEFAULT_CONFIG, resolvePageDimensions } from './config';
import { Theme, mergeTheme } from './theme';
import { PDFDoc, PDFDocumentCtor } from './types';
import { drawTable, TableColumn, TableOptions } from './utils/table';
import { embedImage } from './utils/image';

export interface TemplateData {
  [key: string]: any;
}

export type TemplateRenderer = (
  pdf: PDFDoc,
  config: PDFConfig,
  data: TemplateData,
  theme: Theme
) => void | Promise<void>;

export type { TableColumn, TableOptions };

export class PDFGenerator {
  private doc: PDFDoc;
  private config: PDFConfig;
  private theme: Theme;
  private templates: Map<string, TemplateRenderer> = new Map();

  constructor(config: Partial<PDFConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config, margins: { ...DEFAULT_CONFIG.margins, ...config.margins } };
    this.theme = mergeTheme(this.config.theme);

    const [width, height] = resolvePageDimensions(this.config);

    this.doc = new PDFDocumentCtor({
      size: [width, height],
      margins: this.config.margins,
      bufferPages: true,
      info: { Producer: 'universal-pdf-gen', Creator: 'universal-pdf-gen' }
    });
  }

  registerTemplate(name: string, renderer: TemplateRenderer): void {
    this.templates.set(name, renderer);
  }

  async useTemplate(templateName: string, data: TemplateData): Promise<this> {
    const template = this.templates.get(templateName);
    if (!template) {
      throw new Error(
        `Template "${templateName}" not found. Available: ${Array.from(this.templates.keys()).join(', ') || '(none registered)'}`
      );
    }
    await template(this.doc, this.config, data, this.theme);
    return this;
  }

  /** Usable content width between the left and right margins. */
  get contentWidth(): number {
    return this.doc.page.width - this.config.margins.left - this.config.margins.right;
  }

  get contentHeight(): number {
    return this.doc.page.height - this.config.margins.top - this.config.margins.bottom;
  }

  addText(text: string, options: any = {}): this {
    this.doc.text(text, options);
    return this;
  }

  /** Embeds PNG, JPEG, GIF, WEBP, or SVG (file path or Buffer) at the given top-left position. */
  async addImage(source: string | Buffer, x: number, y: number, width?: number, height?: number): Promise<this> {
    await embedImage(this.doc, source, x, y, { width, height });
    return this;
  }

  /**
   * Renders a table with word-wrapped cells, an optional striped body, and automatic
   * page breaks when the table would run past the bottom margin.
   */
  addTable(columns: TableColumn[], rows: Array<Record<string, any> | any[]>, options: TableOptions = {}): this {
    drawTable(this.doc, this.config, this.theme, columns, rows, options);
    return this;
  }

  /** Adds a new page if the given height would not fit before the bottom margin. */
  ensureSpace(height: number): this {
    if (this.doc.y + height > this.doc.page.height - this.config.margins.bottom) {
      this.doc.addPage();
    }
    return this;
  }

  addPage(): this {
    this.doc.addPage();
    return this;
  }

  addLine(x1: number, y1: number, x2: number, y2: number, color: string = this.theme.border, width = 1): this {
    this.doc.strokeColor(color).lineWidth(width).moveTo(x1, y1).lineTo(x2, y2).stroke();
    return this;
  }

  addRect(x: number, y: number, width: number, height: number, fillColor?: string, radius = 0): this {
    if (radius > 0) {
      this.doc.roundedRect(x, y, width, height, radius);
    } else {
      this.doc.rect(x, y, width, height);
    }
    if (fillColor) {
      this.doc.fill(fillColor);
    } else {
      this.doc.stroke();
    }
    return this;
  }

  setFont(fontName: string, size: number): this {
    this.doc.font(fontName).fontSize(size);
    return this;
  }

  moveDown(amount: number = 1): this {
    this.doc.moveDown(amount);
    return this;
  }

  /** Stamps "Page N of M" at the bottom-right of every page. Call once, right before generate(). */
  addPageNumbers(): this {
    const { doc, theme, config } = this;
    const range = doc.bufferedPageRange();
    for (let i = range.start; i < range.start + range.count; i++) {
      doc.switchToPage(i);
      const pageNum = i - range.start + 1;
      doc
        .font('Helvetica')
        .fontSize(8)
        .fillColor(theme.muted)
        .text(`Page ${pageNum} of ${range.count}`, 0, doc.page.height - config.margins.bottom + 10, {
          width: doc.page.width,
          align: 'center'
        });
    }
    return this;
  }

  async generate(filename: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const stream = fs.createWriteStream(filename);
      this.doc.pipe(stream);
      this.doc.end();

      stream.on('finish', () => resolve());
      stream.on('error', reject);
    });
  }

  /** Resolves with the raw PDF bytes instead of writing to disk — useful for HTTP responses. */
  async generateBuffer(): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      const chunks: Buffer[] = [];
      this.doc.on('data', (chunk: Buffer) => chunks.push(chunk));
      this.doc.on('end', () => resolve(Buffer.concat(chunks)));
      this.doc.on('error', reject);
      this.doc.end();
    });
  }

  getDocument(): PDFDoc {
    return this.doc;
  }

  getTheme(): Theme {
    return this.theme;
  }
}
